File: 깊이, 넓이 우선탐색의 활용/피자 배달 거리.py
import sys
# sys.stdin = open("input.txt", 'r')

def dfs(level, start):
    global result
    if level == m:
        sum = 0
        for i in range(len(house)):
            tmp = 2147000000
            for j in combination:
               tmp = min(tmp, abs(house[i][0] - pizza[j][0]) + abs(house[i][1] - pizza[j][1])) 
            sum += tmp
        if sum < result:
            result = sum
    else:
        for i in range(start, len(pizza)):
            combination[level] = i
            dfs(level + 1, i + 1)

if __name__ == "__main__":
    n, m = map(int, input().split())
    board = [list(map(int, input().split())) for _ in range(n)]

    house = []
    pizza = []
    for i in range(n):
        for j in range(n):
            if board[i][j] == 1:
                house.append((i, j))
            if board[i][j] == 2:
                pizza.append((i, j))
    result = 2147000000
    combination = [0] * m
    dfs(0, 0)
    print(result)

File: 그래프, DFS, BFS 관련 보충문제/피자배달거리.cpp
/*
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

vector<pair<int, int>> house;
vector<pair<int, int>> pizza;
vector<pair<int, int>> result(12);
int mini = 2147000000;
int n, m;
int dis;

void dfs(int level, int s){
	if(level == m){
		int tot = 0;
		for(int i = 0; i < house.size(); i++){
			int tmp = 2147000000;
			for(int j = 0; j < m; j++){
				int dis = abs(house[i].first - result[j].first) + \
						  abs(house[i].second - result[j].second);
				if(dis < tmp) tmp = dis;
			}
			tot += tmp;
		}
		if(tot < mini) mini = tot;
	}
	else{
		for(int i = s + 1; i < pizza.size(); i++){
			result[level] = pizza[i];
			dfs(level + 1, i);
		}
	}
}

int main(void){
	// freopen("input.txt", "r", stdin);
	int i, j, val;
	scanf("%d %d", &n, &m);
	
	for(i = 1; i <= n; i++){
		for(j = 1; j <= n; j++){
			scanf("%d", &val);
			if(val == 1) house.push_back({i, j});
			else if(val == 2) pizza.push_back({i, j});
		}
	}
	
	dfs(0, -1);
	printf("%d", mini);
	
	return 0;
}
*/
#include <stdio.h>
#include <vector>
#include <algorithm>
using namespace std;

vector<pair<int, int>> house;
vector<pair<int, int>> pizza;
vector<pair<int, int>> result(12);
int mini = 2147000000;
int n, m;
int dis;

void dfs(int level, int s){
	if(level == m){
		int tot = 0;
		for(int i = 0; i < house.size(); i++){
			int sum = 2147000000;
			int x1 = house[i].first;
			int y1 = house[i].second;
			for(int j = 0; j < m; j++){
				int x2 = result[j].first;
				int y2 = result[j].second;
				int dis = abs(x1 - x2) + abs(y1 - y2);
				if(dis < sum) sum = dis;
			}
			tot += sum;
		}
		if(tot < mini) mini = tot;
	}
	else{
		for(int i = s; i < pizza.size(); i++){
			result[level] = pizza[i];
			dfs(level + 1, i + 1);
		}
	}
}

int main(void){
	// freopen("input.txt", "r", stdin);
	int i, j, val;
	scanf("%d %d", &n, &m);
	
	for(i = 1; i <= n; i++){
		for(j = 1; j <= n; j++){
			scanf("%d", &val);
			if(val == 1) house.push_back({i, j});
			else if(val == 2) pizza.push_back({i, j});
		}
	}
	
	dfs(0, 0);
	printf("%d", mini);
	
	return 0;
}
